# (실습) 넘파이 어레이

**주의사항**

- `None`은 적절한 표현식으로 대체한다.
- `Pass`는 적절한 명령문으로 대체한다.

## numpy 모듈

**문제 1**

아래 코드를 설명하라.

In [1]:
import numpy as np

`코드설명`: 넘파이 기본 모듈 임폴트

아래 코드는 수정하지 마세요!!!

In [ ]:
np.__version__

**문제 2**

아래 코드의 기능을 설명하라

In [ ]:
np.random.seed(42)

`코드설명`: 42까지의 랜덤시드 생성

**문제 3**

부동소수점으로 구성된 어레이의 항목은 기본적으로 소수점 이하 8자리까지 출력한다.

In [ ]:
Z = np.random.random((40,40))
Z

`np.set_printoptions()` 의 기능을 다음 두 개의 예를 이용하여 설명하라.

In [ ]:
np.set_printoptions(precision=4)

In [ ]:
Z

In [ ]:
np.set_printoptions(precision=6)

In [ ]:
Z

`코드설명`: 어레이에 포함된 부동소수점들의 표현 정확도 지정

아래 코드는 기본 옵션값으로 되돌린다. 수정하지 마세요.

In [ ]:
np.set_printoptions(precision=8)

## 1차원 어레이

**문제 1**

크기가 10이고 모든 항목이 0인 1차원 어레이를
변수 `Z`가 가리키도록 하라.
단, `np.zeros()` 함수를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros(10)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])).all()

**문제 2**

변수 `Z`가 10에서 49까지의 정수 중에서 짝수로만 구성된 1차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(10, 49 ,2)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
                        44, 46, 48])).all()

**문제 3**

변수 `Z`가 아래와 같이 선언되었다.

In [ ]:
Z = np.array([2, 3, 1, 0, 4])
Z

변수 `Z`의 항목을 정렬하라.
단, sort() 메서드를 활용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z.sort()

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.arange(5)).all()

**문제 4**

변수 `Z`가 크기 30이며 무작위로 생성된 1차원 어레이를 가리키도록 하라.
또한 포함된 항목들의 평균값을 계산하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
`np.random.random()` 함수, `mean()` 어레이 메서드를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(30)
Z_mean = Z.mean()

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.shape == (30,))
print(f"1차원 어레이 평균값: {Z_mean}")

**문제 5**

1차원 어레이가 다음과 같이 주어졌다.

In [ ]:
Z = np.arange(1, 4)

(1) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z*Z

`코드설명`: array([1, 2, 3]) * array([1, 2, 3]) = array([1, 4, 9])

(2) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z**Z

`코드설명`: 제곱근

(3) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z < -Z

`코드설명`: array([1, 2, 3]) < array([-1, -2, -3]) = array([False, False, False]

(4) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z/1/1

`코드설명`:  (array([1, 2, 3]) / 1) /1 #동일값 반환

**문제 6**

변수 `Z`가 아래 모양의 어레이를 가리키도록 하라.
단, `np.linspace()` 함수와 `endpoint` 키워드 인자를 이용한다.

```
[0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.linspace(0, 1, 10, endpoint=False)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert np.isclose(Z, np.array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])).all()

**문제 7**

아래 코드의 실행결과를 설명하라.

In [ ]:
Z = np.zeros((10,10))
print("%d bytes" % (Z.size * Z.itemsize))

`코드설명`: 0이 10개로 이루어진 튜플 10개

- Z.size는 10*10=100개
- Z.item은 Z = np.zeros((10,10))이 float64 형 데이터 타입이라 8바이트

**문제 8**

다음 리스트에서 0이 아닌 항목들의 인덱스로 구성된 어레이를 한줄 코드로 생성하라.
단. `np.nonzero()` 함수를 활용한다.

In [ ]:
Z = [1, 2, 0, 0, 4, 0]

In [ ]:
# None을 적절한 표현식으로 대체하라.

non_zeros = np.nonzero(Z)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (non_zeros == np.array([0, 1, 4])).all()

**문제 9**

두 개의 1차원 어레이가 다음과 같이 주어졌다.

In [4]:
A = np.ones(3)*1
B = np.ones(3)*2
print(A)
print(B)

[1. 1. 1.]
[2. 2. 2.]


아래 코드의 실행 결과를 설명하라.

In [3]:
(A+B)*(-A/2)

array([-1.5, -1.5, -1.5])

`코드설명`: A = [1. 1. 1.]
B = [2. 2. 2.]
# - ([3., 3., 3]) * ([-0.5, -0.5, -0.5]) = array([-1.5, -1.5, -1.5])

In [5]:
np.add(A,B,out=B)

array([3., 3., 3.])

`코드설명`: A + B #OUT = B는 기본배열에 결과 저장

In [6]:
np.divide(A,2,out=A)

array([0.5, 0.5, 0.5])

`코드설명`: 나누기

In [7]:
np.negative(A,out=A)

array([-0.5, -0.5, -0.5])

`코드설명`: 부호반전

In [8]:
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

`코드설명`: 요소별 곱셈

## 2차원 어레이

**문제 1**

변수 `Z`가 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [9]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((5, 5))

아래 코드는 수정하지 마세요!!!

In [10]:
assert (Z.shape == (5, 5))
assert (Z.dtype == np.float64)

**문제 2**

변수 `Z`가 0에서 8까지의 정수로 이루어진 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `reshape()` 어레이 메서드를 반드시 이용한다.

In [11]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(9).reshape(3, 3)

아래 코드는 수정하지 마세요!!!

In [12]:
assert (Z == np.array([[0, 1, 2],
                       [3, 4, 5],
                       [6, 7, 8]])).all()

**문제 3**

변수 `Z`가 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, 2차원 어레이의 대각선 상은 1이고 나머지는 0이어야 한다.
또한 `np.eye()` 함수를 반드시 이용한다.

In [13]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [14]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all

**문제 4**

변수 `Z`가 아래와 같은 (5,5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `np.diag()` 함수의 `k` 키워드 인자를 반드시 활용한다.

```
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
```

In [15]:
# None을 적절한 표현식으로 대체하라.
Z = np.diag(1+np.arange(4), k=-1)

아래 코드는 수정하지 마세요!!!

In [16]:
assert (Z == np.array([[0, 0, 0, 0, 0],
                        [1, 0, 0, 0, 0],
                        [0, 2, 0, 0, 0],
                        [0, 0, 3, 0, 0],
                        [0, 0, 0, 4, 0]])).all()

**문제 5**

변수 `Z`가 아래와 같은 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.zeros()` 함수, `np.arange()` 함수, 2차원 어레이 덧셈을 이용한다.

```
[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]
```

In [17]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros((5, 5)) + np.arange(5)

아래 코드는 수정하지 마세요!!!

In [18]:
assert (Z == np.array([[0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.]])).all()

**문제 6**

아래 3x3 단위 행렬에 해당하는 2차원 어레이를 한줄 코드로 생성하라.
단, `np.eye()` 함수를 이용한다.

```
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

In [19]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [20]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all()

**문제 7**

무작위 수로 구성된 (5, 5) 모양의 2차원 어레이가 다음과 같다.

In [21]:
#지정한 범위로 값을 조정하는 것: 정규화
Z = np.random.random((5, 5))

변수 `Z`가 가리키는 2차원 어레이에 대해 min-max 크기조정(정규화)을 적용하라.

min-max 크기조정(정규화)은 항목의 모든 값을 아래 식을 통해 0과 1 사이의 부동소수점으로 변환한다. $max$와 $min$은 각각 최대 항목과 최소 항목을 가리킨다.

$$
\frac{x-min}{max - min}
$$

In [23]:
# None을 적절한 표현식으로 대체하라.
Z = (Z - Z.min()) / (Z.max() - Z.min())

아래 코드는 수정하지 마세요!!!

In [24]:
assert (Z.max() == 1 and Z.min() == 0)

## 3차원 어레이

**문제 1**

(1) 변수 `Z`가 (3, 3, 3) 모양의 3차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [25]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((3, 3, 3))

아래 코드는 수정하지 마세요!!!

In [26]:
assert (Z.shape == (3, 3, 3))
assert (Z.dtype == 'f8')

(2) `Z`에 포함된 항목들 중에서 최댓값과 최솟값의 차이를 구하라.
단, `maximum`, `mininum` 두 변수가 각각 최댓값과 최솟값을 가리키도록 하며,
`max()`와 `min()` 어레이 메서드를 이용한다.

In [27]:
# None을 적절한 표현식으로 대체하라.
maximum = Z.max()
minimum = Z.min()

아래 코드는 수정하지 마세요!!!

In [28]:
print(f"최댓값과 최솟값의 차이: {maximum - minimum}")

최댓값과 최솟값의 차이: 0.9344309403588984


## `dtype` 종류

**문제 1**

아래 두 코드의 실행결과를 설명하라.

In [29]:
for dtype in [np.int8, np.int32, np.int64]:
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)

-128
127
-2147483648
2147483647
-9223372036854775808
9223372036854775807


`코드설명`: 정수형 자료형에 허용되는 비트수에 따른 최소값과 최대값
- 8비트 정수형(np.int8): -128에서 127사이의 정수
- 16비트 정수형: -2147483648에서
2147483647사이의 정수
- 32비트 정수형: -9223372036854775808에서
9223372036854775807 사이의 정수

In [30]:
for dtype in [np.float32, np.float64]:
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
    print(np.finfo(dtype).eps)

-3.4028235e+38
3.4028235e+38
1.1920929e-07
-1.7976931348623157e+308
1.7976931348623157e+308
2.220446049250313e-16


`코드설명`:부동소수점 자료형에 허용되는 비트수에 따른 최소값과 최대값

**문제 2**

부동소수점으로 구성된 1차원 어레이가 아래와 같이 주어졌다.

In [31]:
Z = (np.random.rand(10)*100).astype(np.float32)

변수 `Y`가 변수 `Z`가 가리키는 1차원 어레이의 dtype이 `int32`로 형변환된 1차원 어레이를 가리키도록 하라.
단, `astype()` 메서드를 이용한다.

In [32]:
# None을 적절한 표현식으로 대체하라.
Y = Z.astype(np.int32)

아래 코드는 수정하지 마세요!!!

In [33]:
assert (Y.shape == (10,))
assert (Y.dtype == np.int32)

**문제 3**

변수 `Z`가 아래와 같이 무작위로 생성된 길이가 10인 1차원 어레이를 가리킨다.

In [34]:
Z = np.random.uniform(0,10,10)
Z

array([2.0125949 , 8.60229005, 8.45470152, 9.35388427, 3.89263227,
       4.74516846, 1.82089692, 3.28830607, 5.73129852, 3.74712926])

(1) 나머지 연산자(`%`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [35]:
# None을 적절한 표현식으로 대체하라.
Y = Z - Z%1 #Z의 소수점자리 빼기

아래 코드는 수정하지 마세요!!!

In [36]:
assert (Y == Z.astype(np.int32)).all()

(2) 몫 연산자(`//`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [41]:
# None을 적절한 표현식으로 대체하라.
Y = Z//1

아래 코드는 수정하지 마세요!!!

In [42]:
assert (Y == Z.astype(np.int32)).all()

(3) 내림 함수 `np.floor()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [43]:
# None을 적절한 표현식으로 대체하라.
Y = np.floor(Z)

아래 코드는 수정하지 마세요!!!

In [44]:
assert (Y == Z.astype(np.int32)).all()

(4) `np.trunc()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [45]:
# None을 적절한 표현식으로 대체하라.
Y = np.trunc(Z)

아래 코드는 수정하지 마세요!!!

In [46]:
assert (Y == Z.astype(np.int32)).all()